src/Interpreters/Set.cpp

```diff
@@ -314,10 +379,17 @@ ColumnPtr Set::execute(const ColumnsWithTypeAndName & columns, bool negative) co
     Columns materialized_columns;
     materialized_columns.reserve(num_key_columns);
 
+    /// We will check existence in Set only for keys whose components do not contain any NULL value.
+    ConstNullMapPtr null_map{};
+    ColumnPtr null_map_holder;
+
     for (size_t i = 0; i < num_key_columns; ++i)
     {
         ColumnPtr result;
 
+        null_map = ConstNullMapPtr();
+        null_map_holder = nullptr;
```

Comment:
I think we actually should merge `null_map_holder` together. E.g. we we have `(dt64, dt64) in (dt, dt)` case.